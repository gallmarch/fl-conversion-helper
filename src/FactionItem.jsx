import React from 'react';
import { connect } from 'react-redux';
import { items } from './factions';
import { favoursRequired } from './favours-required';
import { getInventoryMatch } from './Item';
import BlankItem from './BlankItem';
import UsableItem from './UsableItem';
import DummiedItem from './DummiedItem';

function FactionItem(props) {
  const { factions: { favours, renown }, id } = props;
  const faction = items[id];
  const inventoryMatch = getInventoryMatch(id);

  const factionFavours = favours[faction];
  const convertible = (
    factionFavours >= favoursRequired(renown[faction])
  );

  if (!inventoryMatch) {
    return <BlankItem />;
  }

  if (convertible) {
    return (
      <UsableItem
        inventoryMatch={inventoryMatch}
        quantity={factionFavours}
      >
      </UsableItem>
    );
  }
  return <DummiedItem
    inventoryMatch={inventoryMatch}
    quantity={factionFavours}
  />;
}

function mapStateToProps(state) {
  return { factions: state.factions };
}

export default connect(mapStateToProps)(FactionItem);
