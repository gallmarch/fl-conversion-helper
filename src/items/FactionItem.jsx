import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import getFaction from '../factions/getFaction';

import DummiedItem from './DummiedItem';
import UsableItem from './UsableItem';
import { getFailureExplanation, isUsableFaction } from './selectors';

function FactionItem(props) {
  const {
    data,
    element,
    favours,
    isUsable,
  } = props;

  // Show 'X Favours' as the quantity
  const textualLevel = `${favours} Favour${favours === 1 ? '' : 's'}`;

  if (!isUsable) {
    const { failureExplanation } = props;

    return (
      <DummiedItem
        data={{
          ...data,
          textualLevel,
          level: favours,
          secondaryDescription: `${data.secondaryDescription ? data.secondaryDescription : ''}<p><i>${failureExplanation}</i></p>`,
        }}
        element={element}
      />
    );
  }

  return (
    <UsableItem data={{ ...data, textualLevel, level: favours }} element={element} />
  );
}

FactionItem.propTypes = {
  attributes: PropTypes.object.isRequired, // eslint-disable-line
  data: PropTypes.object.isRequired, // eslint-disable-line
  element: PropTypes.object.isRequired, // eslint-disable-line
  failureExplanation: PropTypes.string,
  favours: PropTypes.number.isRequired,
  isUsable: PropTypes.bool.isRequired,
};

FactionItem.defaultProps = {
  failureExplanation: null,
};

const mapStateToProps = (state, props) => ({
  failureExplanation: getFailureExplanation(state, props),
  favours: state.myself.favours[getFaction(props.id)],
  isUsable: isUsableFaction(state, props),
});

export default connect(mapStateToProps)(FactionItem);
