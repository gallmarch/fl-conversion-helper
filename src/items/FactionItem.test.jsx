import React from 'react';
import renderer from 'react-test-renderer';

import { FactionItem } from './FactionItem';

describe('FactionItem', () => {
  describe('when the user has no matching item', () => {
    it('matches its snapshot', () => {
      const props = {
        attributes: { key: 0 },
        enablementPreference: 0,
        factions: { favours: {}, renown: {} },
        id: 'some-component-id',
      };
      const component = renderer.create(<FactionItem {...props} />);
      expect(component).toMatchSnapshot();
    });
  });
});
