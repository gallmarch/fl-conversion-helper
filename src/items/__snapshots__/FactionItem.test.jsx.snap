// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FactionItem when the user has no matching item matches its snapshot 1`] = `
<li
  className="empty-icon flch-empty-icon"
  id={undefined}
/>
`;
